File: Daily Challanges/2026-05-May/2026-05-16.py
def get_longest_chain(dominoes):
    def search(last, used):
        best = []
        for i, d in enumerate(dominoes):
            if i not in used:
                for tile in [d, d[::-1]]:
                    if tile[0] == last:
                        rest = search(tile[1], used | {i})
                        if len(rest) + 1 > len(best):
                            best = [tile] + rest
        return best

    best = []
    for i, d in enumerate(dominoes):
        for tile in [d, d[::-1]]:
            result = [tile] + search(tile[1], {i})
            if len(result) > len(best):
                best = result
    return best

print(get_longest_chain([[1, 2], [4, 5], [2, 3]]))
print(get_longest_chain([[2, 1], [4, 3], [5, 3]]))

File: Daily Challanges/2026-05-May/2026-05-16.js
function getLongestChain(dominoes) {
  function search(last, used) {
    let best = [];
    for (let i = 0; i < dominoes.length; i++) {
      if (!used.has(i)) {
        const d = dominoes[i];
        const options = [d, [d[1], d[0]]];
        for (const tile of options) {
          if (tile[0] === last) {
            const newUsed = new Set(used);
            newUsed.add(i);
            const rest = search(tile[1], newUsed);
            if (rest.length + 1 > best.length) {
              best = [tile, ...rest];
            }
          }
        }
      }
    }

    return best;
  }

  let best = [];

  for (let i = 0; i < dominoes.length; i++) {
    const d = dominoes[i];
    const options = [d, [d[1], d[0]]];

    for (const tile of options) {
      const used = new Set([i]);

      const result = [tile, ...search(tile[1], used)];

      if (result.length > best.length) {
        best = result;
      }
    }
  }

  return best;
}

console.log(
  getLongestChain([
    [1, 2],
    [4, 5],
    [2, 3],
  ]),
);
console.log(
  getLongestChain([
    [2, 1],
    [4, 3],
    [5, 3],
  ]),
);
